hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/ExpressionTestBase.java

```diff
@@ -0,0 +1,1190 @@
+                    break;
+            }
+
+            TYPES.add(new Operand(UNKNOWN_TYPE, type, unparse(type)));
+
+            type = TYPE_FACTORY.createTypeWithNullability(type, true);
+            COLUMNS.add(new Operand(type, UNKNOWN_VALUE, typeName.getName().toLowerCase() + "1"));
+            COLUMNS.add(new Operand(type, UNKNOWN_VALUE, typeName.getName().toLowerCase() + "2"));
+
+            if (typeName == BOOLEAN) {
+                BOOLEAN_COLUMN.add(new Operand(type, UNKNOWN_VALUE, typeName.getName().toLowerCase() + "1"));
+            }
+        }
+
+        // Parameters: one is enough, every occurrence produces a new parameter.
+
+        PARAMETERS.add(new Operand(UNKNOWN_TYPE, UNKNOWN_VALUE, "?"));
+    }
+
+    protected static final List<Operand> ALL = combine(COLUMNS, LITERALS, PARAMETERS);
+
+    private final ExecutorService executor = Executors.newWorkStealingPool(RuntimeAvailableProcessors.get());
+
+    @FunctionalInterface
+    protected interface ExpectedTypes {
+
+        RelDataType[] compute(Operand[] operands);
+
+    }
+
+    @FunctionalInterface
+    protected interface ExpectedValues {
+
+        Object compute(Operand[] operands, RelDataType[] types, Object[] args);
+
+    }
+
+    @After
+    public void after() {
+        // ease the pains of thread deallocation
+        executor.shutdown();
+    }
+
+    @SafeVarargs
+    protected final void verify(SqlOperator operator, ExpectedTypes expectedTypes, ExpectedValues expectedValues,
+                                List<Operand>... operands) {
+        verify(operator, -1, -1, expectedTypes, expectedValues, operands);
+    }
+
+    @SafeVarargs
+    protected final void verify(SqlOperator operator, ExpectedTypes expectedTypes, ExpectedValues expectedValues, String format,
+                                List<Operand>... operands) {
+        verify(operator, -1, -1, expectedTypes, expectedValues, format, operands);
+    }
+
+    @SafeVarargs
+    @SuppressWarnings("SameParameterValue")
+    protected final void verify(SqlOperator operator, int invocationId, int evaluationId, ExpectedTypes expectedTypes,
+                                ExpectedValues expectedValues, List<Operand>... operands) {
+        String format;
+        switch (operator.getSyntax()) {
+            case PREFIX:
+                assert operands.length == 1;
+                format = operator.getName() + "(%s)";
+                break;
+
+            case BINARY:
+                assert operands.length == 2;
+                format = "%s " + operator.getName() + " %s";
+                break;
+
+            default:
+                throw new IllegalArgumentException("unexpected syntax: " + operator.getSyntax());
+        }
+
+        verify(operator, invocationId, evaluationId, expectedTypes, expectedValues, format, operands);
+    }
+
+    @SafeVarargs
+    protected final void verify(SqlOperator operator, int invocationId, int evaluationId, ExpectedTypes expectedTypes,
```

Comment:
Can we propagate ID here to facilitate debugging?

Reply:
It should be propagated already and included in the error messages, if any.

Reply:
Yes, but for the debugging purposes it is harder - you need to add `contains` into the breakpoint expression, instead of just setting `id == 42`.

Reply:
The method already receives and propagates `invocationId` and `evaluationId`, so you can use them in breakpoint expressions. If you need to test some specific combination of `invocationId` and `evaluationId`, you may just pass it directly to `verify` and avoid setting any breakpoints. If `-1` is passed as an ID, that means all possible invocations/evaluations would be verified, the generated IDs are propagated and can be used for breakpoints, if you need to.